backend/ml/mehestan/online_heuristics.py

```diff
@@ -0,0 +1,519 @@
+    * For each case, we compute new global scores for the two entities
+        and we apply poll level scaling at global scores
+
+    """
+    poll = Poll.objects.get(pk=poll_pk)
+    all_comparison_of_user_for_criteria = ml_input.get_comparisons(
+        criteria=criteria, user_id=user_id
+    )
+
+    entity_id_a = Entity.objects.get(uid=uid_a).pk
+    entity_id_b = Entity.objects.get(uid=uid_b).pk
+
+    # For delete comparison, we reintroduce the comparison with a 0 score and treat like an update
+    if delete_comparison_case:
+        df_new_row = pd.DataFrame(
+            [(user_id, entity_id_a, entity_id_b, criteria, np.NaN, 0)],
+            columns=["user_id", "entity_a", "entity_b", "criteria", "score", "weight"],
+        )
+        all_comparison_of_user_for_criteria = pd.concat(
+            [all_comparison_of_user_for_criteria, df_new_row]
+        )
+
+    if not check_requirements_are_good_for_online_heuristics(
+        criteria,
+        all_comparison_of_user_for_criteria,
+        entity_id_a,
+        entity_id_b,
+    ):
+        return
+
+    set_of_entity_to_update = {entity_id_a, entity_id_b}
+    previous_individual_raw_scores = ml_input.get_indiv_score(
+        user_id=user_id, criteria=criteria
+    )
+    previous_individual_raw_scores = previous_individual_raw_scores[
+        ["entity_id", "raw_score"]
+    ]
+    previous_individual_raw_scores = previous_individual_raw_scores.set_index(
+        "entity_id"
+    )
+    new_raw_scores = previous_individual_raw_scores
+    for tau in range(0, TAU_SUBITERATION_NUMBER):
+        if tau > 0:
+            set_of_entity_to_update = compute_and_give_next_set_of_entity_to_update(
+                set_of_entity_to_update, all_comparison_of_user_for_criteria
+            )
+
+        # Now with previous trick and valid condition check,
+        # all_comparison_of_user_for_criteria have a pair (entity_id_b,entity_id_a)
+        new_raw_scores, new_raw_uncertainties = get_new_scores_from_online_update(
+            all_comparison_of_user_for_criteria,
+            set_of_entity_to_update,
+            new_raw_scores,
+        )
+
+    # so far we have recompute new indiv score for a and b and neighbours,
+    # we need to recompute global score for a and b
+    # in order to so, we need all individual scaled score concerning a and b
+    # we will get those and inject the new raw score before scaling for {a|b}/criteria/user_id
+    theta_star_a = new_raw_scores.loc[
+        new_raw_scores.index == entity_id_a, "raw_score"
+    ].values.squeeze()[()]
+    delta_star_a = new_raw_uncertainties.loc[
+        new_raw_uncertainties.index == entity_id_a, "raw_uncertainty"
+    ].values.squeeze()[()]
+    theta_star_b = new_raw_scores.loc[
+        new_raw_scores.index == entity_id_b, "raw_score"
+    ].values.squeeze()[()]
+    delta_star_b = new_raw_uncertainties.loc[
+        new_raw_uncertainties.index == entity_id_b, "raw_uncertainty"
+    ].values.squeeze()[()]
```

Comment:
I don't understand what's going on here. I suppose `.squeeze()` is used to extract the single value. But why `[()]` ?

Reply:
with [()] : it returns  <class 'numpy.float64'>
without : it returns  <class 'numpy.ndarray'>

if you have an much more elegant solution, I will take it for sure than an obscure [()]   :+1:

Reply:
ok, so it could simply be written as 
```py
delta_star_b = new_raw_uncertainties.loc[entity_id_b, "raw_uncertainty"]
```
or I'm missing something?

Reply:
new_raw_uncertainties.loc[entity_id_b, "raw_uncertainty"]  give a series with one or multiple lines (with this syntax, the filter can give multiple rows):

1 entity_a
16    0.717044
Name: raw_uncertainty, dtype: float64 <class 'pandas.core.series.Series'>

 with values : [0.71704443] <class 'numpy.ndarray'>
with values.unsqueeze :  0.71704443 <class 'numpy.ndarray'>
and with values.unsqueeze[()]:  0.71704443 <class 'numpy.float64'>